docs/api/ref/GetAgent.md

```diff
@@ -0,0 +1,235 @@
+
+To handle situations where no Desktop Agent is found, a failover function may be supplied by an app allowing it to start or otherwise connect to a Desktop Agent (e.g. by loading a proprietary adaptor that returns a `DesktopAgent` implementation or by creating a window or iframe of its own that will provide a Desktop Agent Proxy).
+
+The definition of the `getAgent()` function is as follows:
+
+```ts
+type GetAgentType = ( 
+  params?: GetAgentParams,  
+) => Promise<DesktopAgent>; 
+```
+
+A small number of arguments are accepted that can affect the behavior of `getAgent` and to provide a fallback in case a Desktop Agent is not found, allowing the application to start its own agent or use another mechanism (such as a proprietary adaptor) to connect to one.
+
+```ts
+/** 
+ * @typedef {Object} GetAgentParams Type representing parameters passed to the 
+ * getAgent function. 
+ * 
+ * @property {number} timeoutMs Number of milliseconds to allow for an FDC3
+ * implementation to be found before calling the failover function or
+ * rejecting (default 750). Note that the timeout is cancelled as soon as a
+ * Desktop Agent is detected. There may be additional set-up steps to perform
+ * which will happen outside the timeout.
+ * 
+ * @property {string} identityUrl The app's current URL is normally sent to
+ * a web-based desktop agent to help establish its identity. This property
+ * may be used to override the URL sent (to handle situations where an app's
+ * URL is not sufficiently stable to use for identity purposes,  e.g. due to
+ * client-side route changes when navigating within the app). The URL set MUST
+ * match the origin of the application (scheme, hostname, and port) or it will
+ * be ignored. If not specified, the app's current URL will be used.
+ * 
+ * @property {boolean} channelSelector Flag indicating that the application
+ * needs access to a channel selector UI (i.e. because it supports User Channels
+ * and does not implement its own UI for selecting channels). Defaults to 
+ * `true`. MAY be ignored by Desktop Agent Preload (container) implementations.
+ * 
+ * @property {boolean} intentResolver Flag indicating that the application
+ * needs access to an intent resolver UI (i.e. because it supports raising one
+ * or more intents and and does not implement its own UI for selecting target
+ * apps). Defaults to `true`. MAY be ignored by Desktop Agent Preload (container)
+ * implementations. 
+ * 
+ * @property {boolean} dontSetWindowFdc3 For backwards compatibility, `getAgent`
+ * will set a reference to the Desktop Agent implementation at `window.fdc3`
+ * if one does not already exist, and will fire the fdc3Ready event. Defaults to
+ * `false`. Setting this flag to `true` will inhibit that behavior, leaving 
+ * `window.fdc3` unset.
+ * 
+ * @property {function} failover An optional function that provides a
```

Comment:
I think it would be good to make this property a URL or a function.   If a URL, it should basically open a window with that URL, acting in essentially the same way as a window proxy function that you'd end up writing.

Reply:
@robmoffat that is what we had in the original proposal. I believe you asked us to simplify it and remove the URL option. All it changes is who writes and maintains the code lives that loads the window (or iframe...) and then waits for it to be ready.

If we go with the URL option we'll need a clear description of how it will behave (as going that route the app will not ahve any control over properties of the window) .

Reply:
How about offering a util fn or example implementation in the docs to handle it via copy-paste. I'm somewhat agnostic as to what we do at the moment.

I beleive the original proposal was:
`failover?: (args: GetAgentParams) => Promise<URL | WindowProxy | DesktopAgent>`

but then who passes in instances of the URL class? It would therefore be `Promise<string | WindowProxy | DesktopAgent>` and we'd have some error handling to do if the string was not a URL (reject with `AgentError.InvalidFailover `).

Reply:
Ok I was thinking:

`failover?: string | (args: GetAgentParams) => Promise<WindowProxy | DesktopAgent>`

Reply:
Saves a few characters yes. COde working with the string will need to check its a URL, load it up into an iframe and at least wait for the load event to fire before trying to contact it. Thats the same as you need to do for adaptors loaded into an iframe so it could be handled by a util function.

However, we put the current draft to the discussion group for review and the SWG for adoption already. Given this is at the edge of what we're doing it might be better to wait for a subsequent version and focus on releasing what we've done so far.